### PR TITLE
Issue an error when a derived class precedes its base class

### LINKED ISSUE #8633: Playground is not supporting for "wrong order" classes
**Actual behavior:**

I write two classes in the playground and when I run the code I get an error in the console which land on the extend function.

class Snake extends Greeter {
    constructor() {
		super();
	}
	skip(){}
}
class Greeter {
    constructor() {}
	
	skip(){
		console.log('nothig to skip');
		this.preEnd();
		console.log('so just ended');
	}
	preEnd(){}
}

**Expected behavior:**

The order of the class should not matter.

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #8633 and dozens of other reportings of this easily-detected error case

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Why are we only checking between base/derived? Basically any value-side reference to a class should be an error before its declaration.

File: tests/baselines/reference/baseTypeWrappingInstantiationChain.errors.txt [hunk 1]
```diff
@@ -0,0 +1,35 @@
+tests/cases/compiler/baseTypeWrappingInstantiationChain.ts(1,21): error TS2686: A class must be declared after its base class.
+tests/cases/compiler/baseTypeWrappingInstantiationChain.ts(12,25): error TS2686: A class must be declared after its base class.
+
+
+==== tests/cases/compiler/baseTypeWrappingInstantiationChain.ts (2 errors) ====
+    class C<T1> extends CBase<T1> {
```
Comment (sandersn) [inline tests/baselines/reference/baseTypeWrappingInstantiationChain.errors.txt]: Can you reorder these classes to fix the error? I'd rather keep types and symbols around for tests that are not meant to exercise the new error.

Comment (sandersn) [inline tests/baselines/reference/baseTypeWrappingInstantiationChain.errors.txt]: And probably some of the other tests too.

Comment (mhegazy) [inline tests/baselines/reference/baseTypeWrappingInstantiationChain.errors.txt]: +1

**Comment by mhegazy** [comment]
:+1:

**Comment by mihailik** [comment]
Does this cover the ordering across files when used with -outFile? That is, Derived and Base are in different files, but when compiled into the single output the order can go a wrong way — is that also reported?

Also what about this case, will it be (incorrectly) reported as an error?

```typescript
function some() {
  class Derived extends Base {
  }
  
  return new Derived(); 
}

class Base { }
```

Textually Derived preceding the Base, but I think if it's not a static top-level or namespace-level class, the ordering should be disregarded and no warning shown.

**Comment by RyanCavanaugh** [comment]
Yep, this fix uses the function `isBlockScopedNameDeclaredBeforeUse` knows how to handle both `outFile` ordering and "evaluated later" scenarios correctly. It's the same function we use to detect improper ordering / use of `let` variables so that logic should be pretty well ironed-out by now.

**Pull request merged**